src/OrbitQt/orbitmainwindow.cpp

```diff
@@ -1269,21 +1272,88 @@ void OrbitMainWindow::ShowTooltip(std::string_view message) {
+
+  if (source_code_file.open(QIODevice::ReadOnly)) return source_code_file.readAll();
+
+  orbit_source_paths_mapping::MappingManager mapping_manager{};
+  const auto maybe_mapping_file_path = mapping_manager.MapToFirstExistingTarget(file_path);
+  if (maybe_mapping_file_path.has_value()) {
+    source_code_file.setFileName(QString::fromStdString(maybe_mapping_file_path->string()));
+    if (source_code_file.open(QIODevice::ReadOnly)) return source_code_file.readAll();
+  }
+
+  QMessageBox message_box{QMessageBox::Warning, "Source code file not found",
+                          QString("Could not find the source code file \"%1\" on this machine.")
+                              .arg(QString::fromStdString(file_path.string())),
+                          QMessageBox::Cancel, this};
+  auto* pick_file_button = message_box.addButton("Choose file...", QMessageBox::ActionRole);
+
+  constexpr const char* kAutocreateMappingKey = "auto_create_mapping";
+  constexpr const char* kPreviousSourcePathsMappingDirectoryKey =
+      "previous_source_paths_mapping_directory";
+  auto checkbox = std::make_unique<QCheckBox>(
+      "Automatically create a source paths mapping from my chosen file.");
+  checkbox->setToolTip(
+      "If enabled Orbit will automatically try to create a source paths mapping from it. The "
+      "common suffix between the path given in the debug information and the local file path will "
+      "be stripped. From the rest a mapping will be created.");
+  QSettings settings{};
+  checkbox->setCheckState(settings.value(kAutocreateMappingKey, true).toBool() ? Qt::Checked
+                                                                               : Qt::Unchecked);
+  message_box.setCheckBox(checkbox.release());
+
+  QString user_chosen_file;
+  QObject::connect(pick_file_button, &QAbstractButton::clicked, this, [&]() {
+    QDir previous_directory{
+        settings.value(kPreviousSourcePathsMappingDirectoryKey, QDir::currentPath()).toString()};
+    const QString file_name = QString::fromStdString(file_path.filename().string());
+
+    user_chosen_file = QFileDialog::getOpenFileName(
+        this, QString{"Choose %1"}.arg(QString::fromStdString(file_path.string())),
+        previous_directory.filePath(file_name), file_name);
+    message_box.accept();
+  });
+
+  message_box.exec();
+
+  if (user_chosen_file.isEmpty()) return std::nullopt;
+
+  settings.setValue(kAutocreateMappingKey, message_box.checkBox()->isChecked());
+  settings.setValue(kPreviousSourcePathsMappingDirectoryKey, QFileInfo{user_chosen_file}.path());
+  if (message_box.checkBox()->isChecked()) {
+    auto maybe_mapping = orbit_source_paths_mapping::InferMappingFromExample(
+        file_path, std::filesystem::path{user_chosen_file.toStdString()});
+    if (maybe_mapping.has_value()) {
+      orbit_source_paths_mapping::MappingManager mapping_manager{};
+      mapping_manager.AppendMapping(maybe_mapping.value());
```

Comment:
Should we append mapping only if we can open it successfully?

Reply:
I don't have a strong opinion for either ways. So yes, I can do that.

Reply:
Done.